### PR TITLE
chore: drop misleading ANSI-mode incompatibility note from CometSum

### LINKED ISSUE #4074: Fix issues found when documenting current expression incompatibilities
### Describe the bug

Some issues were discovered when reviewing #4067. This issue is a follow up for fixing these.

- `CometSum` incorrectly states that it is incompatible in ANSI mode (https://github.com/apache/datafusion-comet/pull/4067/changes#r3140308388)
- `CometStringRepeat` has an incompatibility that we do not fall back for (see https://github.com/apache/datafusion-comet/pull/4067/changes#r3140328947)
- `CometFromUnixTime` incorrect reports `Incompatible` instead of `Unsupported` for unsupported formats

### Steps to reproduce

_No response_

### Expected behavior

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by coderfender**
Partial fix for #4074

`CometSum` does support ANSI mode but is currently gated at the compat level. This PR essentially reverts it to allow native execution

### CODE REVIEW

File: spark/src/main/scala/org/apache/comet/serde/strings.scala [hunk 1]
```diff
@@ -42,13 +42,27 @@ object CometStringRepeat extends CometExpressionSerde[StringRepeat] {
       inputs: Seq[Attribute],
       binding: Boolean): Option[ExprOuterClass.Expr] = {
     val children = expr.children
+
+    children(1) match {
+      case Literal(count, _) if isNegativeNumber(count) =>
+        withInfo(expr, "Negative repeat count is not supported")
+        return None
+      case _ =>
```
Comment (coderfender) [inline spark/src/main/scala/org/apache/comet/serde/strings.scala]: This will guard against negative repeat character but still goes through for positive counts with an error message . `getIncompatibleReasons()` is to add the right error message on documentation

File: spark/src/main/scala/org/apache/comet/serde/unixtime.scala [hunk 2]
```diff
@@ -29,12 +29,20 @@ import org.apache.comet.serde.QueryPlanSerde.{exprToProtoInternal, optExprWithIn
...
+    "DataFusion's valid timestamp range differs from Spark" +
       " (https://github.com/apache/datafusion/issues/16594)")
 
-  override def getSupportLevel(expr: FromUnixTime): SupportLevel = Incompatible(None)
+  override def getSupportLevel(expr: FromUnixTime): SupportLevel = {
+    if (expr.format == Literal(TimestampFormatter.defaultPattern)) {
+      Incompatible(None)
+    } else {
+      Unsupported(None)
+    }
```
Comment (coderfender) [inline spark/src/main/scala/org/apache/comet/serde/unixtime.scala]: I tried to leverage the right reason to fallback based on the pattern here

File: spark/src/main/scala/org/apache/comet/serde/aggregates.scala [hunk 3]
```diff
@@ -205,8 +205,6 @@ object CometAverage extends CometAggregateExpressionSerde[Average] {
 
 object CometSum extends CometAggregateExpressionSerde[Sum] {
 
-  override def getIncompatibleReasons(): Seq[String] = Seq("Falls back to Spark in ANSI mode.")
-
```
Comment (coderfender) [inline spark/src/main/scala/org/apache/comet/serde/aggregates.scala:213]: we support ansi mode so no need to fallback to spark

File: spark/src/main/scala/org/apache/comet/serde/strings.scala [hunk 4]
```diff
@@ -42,13 +42,27 @@ object CometStringRepeat extends CometExpressionSerde[StringRepeat] {
...
     val rightExpr = exprToProtoInternal(rightCast, inputs, binding)
     val optExpr = scalarFunctionExprToProto("repeat", leftExpr, rightExpr)
     optExprWithInfo(optExpr, expr, leftCast, rightCast)
   }
+
+  private def isNegativeNumber(value: Any): Boolean = value match {
+    case i: Int => i < 0
+    case l: Long => l < 0
+    case _ => false
+  }
```
Comment (coderfender) [inline spark/src/main/scala/org/apache/comet/serde/strings.scala:51]: added this method to be utilized later but can remove it if it is not needed

File: spark/src/main/scala/org/apache/comet/serde/unixtime.scala [hunk 5]
```diff
@@ -29,12 +29,27 @@ import org.apache.comet.serde.QueryPlanSerde.{exprToProtoInternal, optExprWithIn
...
 
-  override def getSupportLevel(expr: FromUnixTime): SupportLevel = Incompatible(None)
+  override def getSupportLevel(expr: FromUnixTime): SupportLevel = {
+    expr.format match {
+      case Literal(fmt, _) =>
+        val formatStr = fmt.toString
+        val defaultPattern = TimestampFormatter.defaultPattern
+        if (formatStr == defaultPattern) {
+          Incompatible(None)
+        } else {
```
Comment (coderfender) [inline spark/src/main/scala/org/apache/comet/serde/unixtime.scala]: fallback based on specific pattern

File: spark/src/test/resources/sql-tests/expressions/datetime/from_unix_time.sql [hunk 6]
```diff
@@ -24,12 +24,12 @@ INSERT INTO test_from_unix_time VALUES (0), (1718451045), (-1), (NULL), (2147483
 query expect_fallback(not fully compatible with Spark)
 SELECT from_unixtime(t) FROM test_from_unix_time
 
-query expect_fallback(not fully compatible with Spark)
+query expect_fallback(Datetime pattern format: yyyy-MM-dd is unsupported)
 SELECT from_unixtime(t, 'yyyy-MM-dd') FROM test_from_unix_time
```
Comment (coderfender) [inline spark/src/test/resources/sql-tests/expressions/datetime/from_unix_time.sql:28]: error / fallback reason is more actionable

**Review by coderfender — COMMENTED** [review]
added some comments to help with review

File: spark/src/main/scala/org/apache/comet/serde/strings.scala [hunk 7]
```diff
@@ -33,7 +33,7 @@ import org.apache.comet.serde.QueryPlanSerde.{createBinaryExpr, exprToProtoInter
 
 object CometStringRepeat extends CometExpressionSerde[StringRepeat] {
 
-  override def getCompatibleNotes(): Seq[String] = Seq(
+  override def getIncompatibleReasons(): Seq[String] = Seq(
```
Comment (andygrove) [inline spark/src/main/scala/org/apache/comet/serde/strings.scala]: There is an approved PR to fix the behavior here 

https://github.com/apache/datafusion-comet/pull/4017

File: spark/src/main/scala/org/apache/comet/serde/unixtime.scala [hunk 8]
```diff
@@ -29,12 +29,27 @@ import org.apache.comet.serde.QueryPlanSerde.{exprToProtoInternal, optExprWithIn
...
-    "Only supports the default datetime format pattern `yyyy-MM-dd HH:mm:ss`." +
-      " DataFusion's valid timestamp range differs from Spark" +
+    "DataFusion's valid timestamp range differs from Spark" +
       " (https://github.com/apache/datafusion/issues/16594)")
 
-  override def getSupportLevel(expr: FromUnixTime): SupportLevel = Incompatible(None)
+  override def getSupportLevel(expr: FromUnixTime): SupportLevel = {
+    expr.format match {
+      case Literal(fmt, _) =>
+        val formatStr = fmt.toString
```
Comment (andygrove) [inline spark/src/main/scala/org/apache/comet/serde/unixtime.scala]: Possible NPE if `fmt` is null

File: spark/src/main/scala/org/apache/comet/serde/strings.scala (hunk 7 above)
Comment (andygrove) [reply spark/src/main/scala/org/apache/comet/serde/strings.scala]: This PR has now been merged

Comment (coderfender) [reply spark/src/main/scala/org/apache/comet/serde/strings.scala]: thank you . I removed this check alltogether

File: spark/src/main/scala/org/apache/comet/serde/strings.scala (hunk 1 above)
Comment (coderfender) [reply spark/src/main/scala/org/apache/comet/serde/strings.scala]: no longer relevant

File: spark/src/main/scala/org/apache/comet/serde/strings.scala (hunk 7 above)
Comment (andygrove) [inline spark/src/main/scala/org/apache/comet/serde/strings.scala]: `getIncompatibleReasons` needs to be removed now that this issue has been fixed in https://github.com/apache/datafusion-comet/pull/4017

Comment (andygrove) [reply spark/src/main/scala/org/apache/comet/serde/strings.scala]: Actually that was the wrong issue ... why does this PR rename this method? There is no `getSupportLevel` so this expression is not marked as incompatible

File: spark/src/main/scala/org/apache/comet/serde/unixtime.scala [hunk 9]
```diff
@@ -29,12 +29,27 @@ import org.apache.comet.serde.QueryPlanSerde.{exprToProtoInternal, optExprWithIn
...
+        val defaultPattern = TimestampFormatter.defaultPattern
+        if (formatStr == defaultPattern) {
+          Incompatible(None)
+        } else {
+          Unsupported(Some(s"Datetime pattern format: $formatStr is unsupported"))
+        }
+      case _ =>
+        Unsupported(Some("Datetime pattern format is unsupported"))
```
Comment (andygrove) [inline spark/src/main/scala/org/apache/comet/serde/unixtime.scala]: Message could be clearer by stating that non-literal formats are not supported?

File: spark/src/main/scala/org/apache/comet/serde/unixtime.scala [hunk 10]
```diff
@@ -48,17 +63,18 @@ object CometFromUnixTime extends CometExpressionSerde[FromUnixTime] {
...
-      None
+    expr.format match {
+      case Literal(fmt, _) if fmt != null && fmt.toString != TimestampFormatter.defaultPattern =>
+        withInfo(expr, "Datetime pattern format is unsupported")
+        None
+      case _ if secExpr.isDefined && formatExpr.isDefined =>
+        val timestampExpr =
+          scalarFunctionExprToProto("from_unixtime", Seq(secExpr, timeZone): _*)
+        val optExpr = scalarFunctionExprToProto("to_char", Seq(timestampExpr, formatExpr): _*)
+        optExprWithInfo(optExpr, expr, expr.sec, expr.format)
```
Comment (andygrove) [inline spark/src/main/scala/org/apache/comet/serde/unixtime.scala]: is this code still reachable?

File: spark/src/main/scala/org/apache/comet/serde/strings.scala (hunk 7 above)
Comment (andygrove) [reply spark/src/main/scala/org/apache/comet/serde/strings.scala]: Sorry, I have caused confusion here. #4017 was for substring not string repeat. I do not think that this PR should be touching `CometStringRepeat`? The PR title is `fix spark ansi sum incompatibility message` and this seems unrelated.

Comment (coderfender) [reply spark/src/main/scala/org/apache/comet/serde/strings.scala]: True ! Thank you for pointing this out. I fixed it in the latest commit

File: spark/src/main/scala/org/apache/comet/serde/unixtime.scala (hunk 9 above)
Comment (coderfender) [reply spark/src/main/scala/org/apache/comet/serde/unixtime.scala]: thank you for the idea ! I updated the message to reflect that only non-literal formats are not supported

File: spark/src/main/scala/org/apache/comet/serde/unixtime.scala (hunk 10 above)
Comment (coderfender) [reply spark/src/main/scala/org/apache/comet/serde/unixtime.scala]: thank you . this code is no longer reachable given the compatibility checks `getSupportLevel`

File: spark/src/main/scala/org/apache/comet/serde/strings.scala (hunk 7 above)
Comment (andygrove) [reply spark/src/main/scala/org/apache/comet/serde/strings.scala]: some comments got lost I think due to GitHub outages ... I don't understand why this PR is removing getCompatibleNotes and adding getIncompatibleReasons

**Review by andygrove — CHANGES_REQUESTED** [review]
I want to make sure this PR does not get merged accidentally before feedback is addressed

File: spark/src/test/resources/sql-tests/expressions/string/string_repeat.sql [hunk 11]
```diff
@@ -35,3 +35,11 @@ SELECT repeat('hi', n) FROM test_repeat
 -- literal + literal
 query
 SELECT repeat('hi', 3), repeat('', 5), repeat('a', 0), repeat(NULL, 3)
+
+-- non-positive literal count
```
Comment (andygrove) [inline spark/src/test/resources/sql-tests/expressions/string/string_repeat.sql]: can you add non-literal test as well - that does not seem to be covered in the code change

**Comment by andygrove** [comment]
@coderfender This PR has the title `fix spark ansi sum incompatibility message` but also makes changes to `string_repeat` and `from_unix_time`.  There is also no information in the PR description. Maybe you could update the title and description?

**Comment by coderfender** [comment]
Seems like negative repeat edge case is fixed in DF v53 ? https://github.com/apache/datafusion/blame/4a41173ba3df9b5d47638599c819a1e6e46ad92b/datafusion/functions/src/string/repeat.rs#L165

File: spark/src/main/scala/org/apache/comet/serde/unixtime.scala [hunk 12]
```diff
@@ -29,17 +30,40 @@ import org.apache.comet.serde.QueryPlanSerde.{exprToProtoInternal, optExprWithIn
...
-  override def getSupportLevel(expr: FromUnixTime): SupportLevel = Incompatible(None)
+  override def getSupportLevel(expr: FromUnixTime): SupportLevel = {
+    expr.format match {
+      case Literal(null, _) =>
+        Compatible(None)
+      case Literal(fmt, _) =>
+        val formatStr = fmt.toString
+        val defaultPattern = TimestampFormatter.defaultPattern
+        if (formatStr == defaultPattern) {
+          Incompatible(None)
```
Comment (andygrove) [inline spark/src/main/scala/org/apache/comet/serde/unixtime.scala]: This should be `Incompatible(Some(reason))` so that the plan is annotated

File: spark/src/main/scala/org/apache/comet/serde/strings.scala [hunk 13]
```diff
@@ -33,10 +33,6 @@ import org.apache.comet.serde.QueryPlanSerde.{createBinaryExpr, exprToProtoInter
 
 object CometStringRepeat extends CometExpressionSerde[StringRepeat] {
 
-  override def getCompatibleNotes(): Seq[String] = Seq(
```
Comment (andygrove) [inline spark/src/main/scala/org/apache/comet/serde/strings.scala:36]: This does not appear to have been fixed yet.

https://github.com/apache/datafusion-comet/issues/4462

Comment (coderfender) [reply spark/src/main/scala/org/apache/comet/serde/strings.scala:36]: Interesting.. I did add a test to make sure we are checking for negative repeat count and it seems to be passing : https://github.com/apache/datafusion-comet/pull/4111/changes#diff-ac46886742f037b94ca64e4c0422494e7fef8ea954074178ccd9d4398d61980fR41

**Review by andygrove — APPROVED** [review]
LGTM. Thanks @coderfender.
